### PR TITLE
Improve docs

### PR DESCRIPTION
**Pull request by dcermak**

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -17,46 +17,49 @@ Pytest container
...
-<https://testinfra.readthedocs.io/en/latest/>`_. Assuming you want to
-automatically spin up a container for a test, then the `container` fixture will
-do exactly that (plus it will cleanup after itself):
+``pytest_container`` is a small and simple `pytest <https://pytest.org>`_ plugin
+to test container images via pytest fixtures and `testinfra
+<https://testinfra.readthedocs.io/en/latest/>`_. It takes care of all the boring
+tasks, like spinning up containers, finding free ports and cleaning up after
+tests, and allows you to focus on implementing the actual tests.
+
+Some of the features include:
```
Comment (dmpop) [inline README.rst:26]: These are not features but tasks the plugin performs, no? In this case, I'd suggest something like this:

```suggestion
The plugin automates the following tasks:
```

File: README.rst [hunk 2]
```diff
@@ -17,46 +17,49 @@ Pytest container
...
+- build containers via a :file:`Dockerfile`
+- wait for containers to become healthy before executing tests
+- bind exposed container ports to free ports on the host
+- mount volumes via temporary directories
+- parallel test execution via pytest-xdist
+- dependency between container images
+- run the same test on as many container images as you'd like
+
+At its heart, ``pytest_container`` provides four fixtures to help you achieve
+everything that you need for your container testing. Spinning up a container
```
Comment (dmpop) [inline README.rst:37]:
```suggestion
``pytest_container`` provides four fixtures that give you everything you need for testing containers. Spinning up a container
```

File: README.rst [hunk 3]
```diff
@@ -17,46 +17,49 @@ Pytest container
...
-the test function itself, we can leverage testinfra to run some basic tests
-inside the container itself, e.g. check whether files are there, packages are
-installed, etc.pp.
+The fixture will automatically pull and spin up the container for you, stop it
+and remove it after the test ran. Your test function will receive an instance of
+:py:class:`pytest_container.container.ContainerData`, which includes besides the
+container's id and the class instance which created it also a `testinfra
+<https://testinfra.readthedocs.io/en/latest/>`_ connection. This connection can
+be used to run some basic tests inside the container itself, e.g. check whether
+files are there, packages are installed, etc.pp.
```
Comment (dmpop) [inline README.rst:57]:
```suggestion
The fixture automatically pulls and spins up the container, stops it
and removes it after the test is completed. Your test function receives an instance of
:py:class:`pytest_container.container.ContainerData` that includes besides the
container's id and the class instance which created it also a `testinfra
<https://testinfra.readthedocs.io/en/latest/>`_ connection. This connection can
be used to run basic tests inside the container itself. For example, you can check whether
files are present, packages are installed, etc.
```

> Your test function receives an instance of
:py:class:`pytest_container.container.ContainerData` that includes besides the
container's id and the class instance which created it also a `testinfra
<https://testinfra.readthedocs.io/en/latest/>`_ connection

I don't understand this sentence.

File: source/tutorials.rst [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+=========
+
+
+Getting started with ``pytest_container``
+-----------------------------------------
+
+.. note::
+
+   In this tutorial we will start with a minimal python project and start using
+   ``pytest_container`` to test its command line interface.
```
Comment (dmpop) [inline source/tutorials.rst:11]: _The following tutorial demonstrates how to_ would be better methinks.

- What does _its_ refer to?
- Command-line interface

File: source/tutorials.rst [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+
+We will use `poetry <https://python-poetry.org/>`_ to manage our python
+dependencies in this example and assume that you already have a python project
+with a :file:`pyproject.toml`. If not, run :command:`poetry init` to create a
+new one interactively.
```
Comment (dmpop) [inline source/tutorials.rst:17]:
```suggestion
The example uses  `poetry <https://python-poetry.org/>`_ to manage the Python
dependencies. If the :file:`pyproject.toml` doesn't already exist, create it using the :command:`poetry init`  command.
```

File: source/tutorials.rst [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+First, we need to add ``pytest_container`` as a development dependency:
```
Comment (dmpop) [inline source/tutorials.rst:19]:
```suggestion
Start by adding ``pytest_container`` as a development dependency:
```

File: source/tutorials.rst [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+.. code-block:: shell-session
+
+   ❯ poetry add --dev pytest_container
+
+
+We also recommend to add `pytest-xdist
+<https://github.com/pytest-dev/pytest-xdist>`_ so that tests can be executed in
+parallel:
```
Comment (dmpop) [inline source/tutorials.rst:28]:
```suggestion
Add `pytest-xdist
<https://github.com/pytest-dev/pytest-xdist>`_, so that tests can be executed in
parallel:
```

File: source/tutorials.rst [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+.. code-block:: shell-session
+
+   ❯ poetry add --dev pytest-xdist
+
+
+Now we are ready to start writing our tests. Start by creating (if not yet
+present) a directory for your tests, preferably called :file:`tests` or
+:file:`test`. Add an empty :file:`tests/__init__.py` and create the file
+:file:`tests/conftest.py` with the following contents:
```
Comment (dmpop) [inline source/tutorials.rst:38]:
```suggestion
To keep things tidy, create a directory called :file:`tests`  for your tests. Add an empty file called :file:`tests/__init__.py`. Create the :file:`tests/conftest.py` file with the following contents:
```

File: source/tutorials.rst [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+
+   def pytest_generate_tests(metafunc):
+       auto_container_parametrize(metafunc)
+
+
+
+The above code snippet ensures that the ``auto_container`` and
+``auto_container_per_test`` fixtures work correctly (more specifically, that
+they are using the images defined in ``CONTAINER_IMAGES``).
```
Comment (dmpop) [inline source/tutorials.rst:52]:
```suggestion
The above code snippet ensures that the ``auto_container`` and
``auto_container_per_test`` fixtures work correctly (that is,
they use the images defined in ``CONTAINER_IMAGES``).
```

File: source/tutorials.rst [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+   FROM registry.opensuse.org/opensuse/tumbleweed
+
+   RUN zypper -n in python3-pip
+   COPY dist/*whl .
+   RUN pip install *whl
+
+
+Assuming that this :file:`Dockerfile` were in our project's root and our python
+project is built into the :file:`dist/` folder, we would build an image based on
+openSUSE Tumbleweed with our project installed into it.
```
Comment (dmpop) [inline source/tutorials.rst:71]:
```suggestion
Assuming that the :file:`Dockerfile` resides in the project's root and the Python
project is stored in the :file:`dist/` folder, build an image based on
openSUSE Tumbleweed with the project installed into it.
```

This still doesn't make sense to me. Project installed into what (what does _into it_ refer to)? Can project be installed into something?

File: source/tutorials.rst [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+We can plug this :file:`Dockerfile` into ``pytest_container`` using the
+:py:class:`~pytest_container.container.DerivedContainer` class as follows:
```
Comment (dmpop) [inline source/tutorials.rst:74]:
```suggestion
Plug the :file:`Dockerfile` into ``pytest_container`` using the
:py:class:`~pytest_container.container.DerivedContainer` class as follows:
```

File: source/tutorials.rst [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+       containerfile=dedent("""
+       RUN zypper -n in python3-pip
+       COPY dist/*whl .
+       RUN pip install *whl
+       """)
+   )
+
+Note that we omitted the ``FROM`` line from the
+:py:attr:`~pytest_container.container.DerivedContainer.containerfile` parameter
+as ``pytest_container`` will automatically include it for us.
```
Comment (dmpop) [inline source/tutorials.rst:93]:
```suggestion
Note that  the ``FROM`` line is omitted from the
:py:attr:`~pytest_container.container.DerivedContainer.containerfile` parameter,
as ``pytest_container`` includes it automatically.
```

File: source/tutorials.rst [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+
+Now we add the above snippet into a new file :file:`tests/test_cli.py` and add
+the following test function and global variable as well:
```
Comment (dmpop) [inline source/tutorials.rst:96]:
```suggestion
Add the above snippet to a new file :file:`tests/test_cli.py`, and add
the following test function along with a global variable:
```

File: source/tutorials.rst [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+       res = auto_container.connection.run_expect([0], "my-binary --help")
+       assert "My cool project" in res.stdout
+
+
+The global variable ``CONTAINER_IMAGES`` instructs ``pytest_container`` to run
+all test functions that use the ``auto_container`` or
+``auto_container_per_test`` fixtures once for each image defined in that
+list. This allows you to have a single file with a plethora of tests that shall
+all be executed inside multiple container images and thus saves us the burden of
+manually parametrizing each of them.
```
Comment (dmpop) [inline source/tutorials.rst:112]:
```suggestion
The global variable ``CONTAINER_IMAGES`` instructs ``pytest_container`` to run
all test functions that use the ``auto_container`` or
``auto_container_per_test`` fixtures once for each image defined in that
list. This allows you to have a single file with multiple of tests that to be executed inside multiple container images, thus avoiding the task of parametrizing each of the test manually.
```

File: source/tutorials.rst [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+
+The test function receives a
+:py:class:`~pytest_container.container.ContainerData` instance, where we then use
+the :py:attr:`~pytest_container.container.ContainerData.connection` attribute to
+execute our projects binary. We leverage the `run_expect
+<https://testinfra.readthedocs.io/en/latest/modules.html#testinfra.host.Host.run_expect>`_
+function to execute the binary and check that its exit code is
+``0``. Afterwards, we assert that a search string is in the standard output.
```
Comment (dmpop) [inline source/tutorials.rst:120]:
```suggestion
The test function receives a
:py:class:`~pytest_container.container.ContainerData` instance, where 
the :py:attr:`~pytest_container.container.ContainerData.connection` attribute is used to
execute the project's binary. The `run_expect
<https://testinfra.readthedocs.io/en/latest/modules.html#testinfra.host.Host.run_expect>`_
function is used to execute the binary and check that its exit code is
``0``. Afterwards, we assert that a search string is in the standard output.
```
_leverage_ is a stupid word.

> Afterwards, we assert that a search string is in the standard output.

Assert a search  string? What does this mean?

File: source/tutorials.rst [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+
+Now we can execute this test via :command:`poetry run pytest`.
```
Comment (dmpop) [inline source/tutorials.rst:122]:
```suggestion
You can now execute this test via :command:`poetry run pytest`.
```

File: source/tutorials.rst (hunk 8 above)
Comment (dcermak) [reply source/tutorials.rst:38]: :broom:

File: README.rst [hunk 17]
```diff
@@ -17,46 +17,49 @@ Pytest container
...
-automatically spin up a container for a test, then the `container` fixture will
-do exactly that (plus it will cleanup after itself):
+``pytest_container`` is a `pytest <https://pytest.org>`_ plugin
+to test container images via pytest fixtures and `testinfra
+<https://testinfra.readthedocs.io/en/latest/>`_. It takes care of all the boring
+tasks, like spinning up containers, finding free ports and cleaning up after
+tests, and allows you to focus on implementing the actual tests.
+
+The plugin automates the following tasks:
+- automatically pull, launch & stop containers
```
Comment (dmpop) [inline README.rst:27]:
```suggestion
- pull, launch, and stop containers
```

File: README.rst [hunk 18]
```diff
@@ -17,46 +17,49 @@ Pytest container
...
+- build containers via a :file:`Dockerfile`
```
Comment (dmpop) [inline README.rst:28]:
```suggestion
- build containers using a :file:`Dockerfile`
```

File: README.rst [hunk 19]
```diff
@@ -17,46 +17,49 @@ Pytest container
...
+- wait for containers to become healthy before executing tests
+- bind exposed container ports to free ports on the host
+- mount volumes via temporary directories
+- parallel test execution via pytest-xdist
```
Comment (dmpop) [inline README.rst:32]:
```suggestion
- parallel test execution through pytest-xdist
```

File: README.rst [hunk 20]
```diff
@@ -17,46 +17,49 @@ Pytest container
...
+- dependency between container images
```
Comment (dmpop) [inline README.rst:33]: I feel like there is something missing. dependency resolution, management?

File: README.rst [hunk 21]
```diff
@@ -17,46 +17,49 @@ Pytest container
...
+- run the same test on as many container images as you'd like
```
Comment (dmpop) [inline README.rst:34]:
```suggestion
- run the same test on as many container images as necessary
```

File: source/tutorials.rst [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+=========
+
+
+Getting started with ``pytest_container``
+-----------------------------------------
+
+.. note::
+
+   The following tutorial demonstrates how to get started using
+   ``pytest_container`` for a python project and to test the project's
```
Comment (dmpop) [inline source/tutorials.rst:11]:
```suggestion
   The following tutorial demonstrates how to use
   ``pytest_container`` for a Python project and to test the project's
```

File: source/tutorials.rst [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+Add `pytest-xdist
+<https://github.com/pytest-dev/pytest-xdist>`_, so that tests can be executed in
+parallel:
+
+.. code-block:: shell-session
+
+   ❯ poetry add --dev pytest-xdist
+
+
+To keep things tidy, create a directory called :file:`tests`  for your tests. Add an empty file called :file:`tests/__init__.py`. Create the :file:`tests/conftest.py` file with the following contents:
```
Comment (grisu48) [inline source/tutorials.rst:34]: "To keep things tidy" is a bit too abstract IMHO because it states an intention but not the effect. I would rephrase it as follows to put emphasis, that it's good to keep the tests atop of this framework in one single place, and that this place can be arbitrary. e.g.

```suggestion
It's highly recommended to create a new directory e.g. :file:`tests`  for your tests to keep things tidy. Add an empty file called :file:`tests/__init__.py`. Create the :file:`tests/conftest.py` file with the following contents:
```

**Review by grisu48 — APPROVED** [review]
Documentation improvements are always a good idea! I like the changes and have only one minor suggestion to add.

Overall LGTM!

File: README.rst (hunk 20 above)
Comment (dcermak) [reply README.rst:33]: Hm, it doesn't really do dependency resolution. All it can do is the following: image B needs A to build, so it will first build A and then B (you can do that recursively as well). But there's no real dependency resolution here (and it will certainly fail if you introduce cycles)

**Review by grisu48 — APPROVED** [review]
